src/base/Notifier.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {ISubscriber} from "../interfaces/ISubscriber.sol";
+import {PositionConfig} from "../libraries/PositionConfig.sol";
+
+contract Notifier {
+    error SubscriberCannotBeNotified();
+    error AlreadySubscribed(address subscriber);
+
+    event Subscribed(uint256 tokenId, address subscriber);
+    event Unsubscribed(uint256 tokenId, address subscriber);
+
+    uint256 private immutable subscriberGasLimit;
+    ISubscriber private constant NO_SUBSCRIBER = ISubscriber(address(0));
+
+    mapping(uint256 tokenId => ISubscriber subscriber) public subscriber;
+
+    constructor(uint256 _subscriberGasLimit) {
+        subscriberGasLimit = _subscriberGasLimit;
+    }
+
+    function _subscribe(uint256 tokenId, PositionConfig memory config, address newSubscriber) internal {
+        ISubscriber _subscriber = subscriber[tokenId];
+
+        if (_subscriber != NO_SUBSCRIBER) revert AlreadySubscribed(address(_subscriber));
+        subscriber[tokenId] = ISubscriber(newSubscriber);
+
+        ISubscriber(newSubscriber).notifySubscribe(tokenId, config);
+        emit Subscribed(tokenId, address(newSubscriber));
+    }
+
+    /// @dev Must always allow a user to unsubscribe. In the case of a malicious subscriber, a user can always unsubscribe safely, ensuring liquidity is always modifiable.
+    function _unsubscribe(uint256 tokenId, PositionConfig memory config) internal {
+        ISubscriber _subscriber = subscriber[tokenId];
+
+        if (gasleft() < subscriberGasLimit) revert SubscriberCannotBeNotified();
+        try _subscriber.notifyUnsubscribe{gas: subscriberGasLimit}(tokenId, config) {} catch {}
```

Comment:
is this subject to return data bombing? i feel like yes? we might have to do this call in assembly to avoid it 👀

Reply:
i think no, because we call it in solidity and it expects to return data (spoke about this in the thread with daniel)

Reply:
ok, worth writing a test for one with a huge return data then to show that that isnt true